Fetch MNTOHA data from ScienceBase using Snakemake

AndyMcAliley opened the pull request:
This is a first pass at `1_fetch`, using MNTOHA since that download is straightforward. Eventually we'll use more lakes than MNTOHA. This uses Snakemake to manage the pipeline. Closes #1 and closes #4. #2 is addressed except for ensuring that pytorch is gpu-enabled.

# `Snakemake`

This is my first time using `Snakemake`, so I'm probably not using it optimally. Still, here's a walkthrough of how it's set up. I set this up while consulting the [docs](https://snakemake.readthedocs.io/en/stable/index.html), [drb-estuary-salinity-ml](https://github.com/USGS-R/drb-estuary-salinity-ml), and [river-dl](https://github.com/USGS-R/river-dl).

The Snakefile contains all the rules for what needs to be made. The first rule is always the default rule that gets made if you call `snakemake` from the command line without specifying any particular rule. So, generally, that rule is the "all" rule. Here, it's called `fetch_all`. It ensures that every file we want is, in fact, downloaded in the proper place. It does that by requiring all those files as inputs. This rule alone has no outputs because it relies on other rules to actually do the downloading - it just lists all the files we want as dependencies. It also has one other input: `1_fetch/in/pull_date.txt`. This file just lists a date. It's a dummy file that can be altered when we want to trigger a fresh download.

The way `fetch_all` requires all the files we want is by referencing the config dictionary, which is loaded from `1_fetch/fetch_config.yaml`. There, you'll find every individual file listed, plus ScienceBase IDs.

Running `snakemake -c1 -p fetch_all` specifies this rule, which downloads everything.

The other rules are "[generalized](https://snakemake.readthedocs.io/en/stable/tutorial/basics.html#step-2-generalizing-the-read-mapping-rule)" rules for how to download files with file paths that match a specific pattern. For instance, say we want to download `1_fetch/out/metadata/lake_metadata.csv`. `Snakemake` notices that this file path matches the pattern `1_fetch/out/metadata/*`, which is the pattern specified in the output of the `fetch_metadata` rule. So, `Snakemake` uses the `fetch_metadata` rule for that file. Likewise, The `fetch_obs` rule is for any files downloaded to `1_fetch/out/obs/`, and the `fetch_driver` rule is for any files downloaded to `1_fetch/out/driver/`.

The `shell` directive is followed by the shell command to run. That command is expected to produce the output files in the `output` directive. Here, the `shell` commands for the three rules all call the same python script, `1_fetch/src/sb_fetch.py`, but with different arguments. The first argument is the ScienceBase ID, and the second argument is the file path. The ScienceBase ID is drawn from the `Snakemake` config file, found at `1_fetch/fetch_config.yaml`. Each rule has its own ScienceBase ID associated with it. 

This aspect of the pipeline is a bit brittle. Since the download path is used to determine which rule to use, and each rule is associated with one SB ID, that means that every SB ID has to have its own subdirectory in `1_fetch/out/`. Not a huge deal, but I suspect there's a more robust way. It may not matter, though, since we'll eventually move away from MNTOHA and probably SB downloads. I do want to get my head around `Snakemake` best practices, though.

The good news is that this method makes `Snakemake` aware of every file to download. This way, files can be downloaded in parallel. If one file is missing or the download failed partway through, calling `Snakemake` again will cause it to pick up where it left off, without needing to re-download everything from that SB ID.

jsadler2 commented on 1_fetch/fetch_config.yaml:
```diff
@@ -0,0 +1,68 @@
+sb_fetch.py:
```
I would consider restructuring this config file with another level, the category. (`metadata`, `obs`,  `drivers`). 

See example here: https://github.com/jsadler2/lake-temperature-lstm-static/blob/snakefile-shorter/1_fetch/fetch_config.yaml

jsadler2 replied:
This allows you to access the `sb_id` more easily: https://github.com/jsadler2/lake-temperature-lstm-static/blob/ef32613bf46a67df06a80195472ac4518718be28/Snakefile#L24

AndyMcAliley replied:
Done! This structure does make things easier.

jsadler2 commented on 1_fetch/fetch_config.yaml:
```diff
@@ -0,0 +1,68 @@
+  driver_id: 5e5d0b96e4b01d50924f2b34
+  driver_files:
+    - 1_fetch/out/driver/04_inputs.xml
+    - 1_fetch/out/driver/clarity_01_N48.00-49.50_W89.50-97.25.zip
```
If you use `expand` and separate out the "driver types" (`clarity`, `ice_flags`, `inputs`) you can let snakemake come up with all of the combinations of these.

* `expand` example: https://github.com/jsadler2/lake-temperature-lstm-static/blob/ef32613bf46a67df06a80195472ac4518718be28/Snakefile#L13

* breaking out "driver types" example: https://github.com/jsadler2/lake-temperature-lstm-static/blob/ef32613bf46a67df06a80195472ac4518718be28/1_fetch/fetch_config.yaml#L18

AndyMcAliley replied:
I like how short the config file is with this change. However, rather than let Snakemake create the combinations, I moved that logic into the `get_input_files` function (now renamed as `get_mntoha_input_files`). That feels a bit easier to read and maintain to me?

jsadler2 commented on README.md:
````diff
@@ -1,2 +1,38 @@
+Note: if [Mamba](https://github.com/mamba-org/mamba) is available, the environment may be created significantly faster by running
+
+```
+mamba update -n base mamba
````
I didn't know about `mamba`. That looks helpful!

AndyMcAliley replied:
I just started using `mamba`. It was the `Snakemake` docs that pointed me to it. It is quick, especially when it comes to resolving package conflicts, and it's been easy to use as a drop-in replacement for `conda` in most circumstances.

jsadler2 commented on Snakefile:
```diff
@@ -0,0 +1,37 @@
+        expand("{file}", file=config["sb_fetch.py"]["driver_files"])
+
+# Lake information for 881 MNTOHA lakes
+rule fetch_metadata:
```
with restructuring the config file, you could use just one rule (see https://github.com/jsadler2/lake-temperature-lstm-static/blob/ef32613bf46a67df06a80195472ac4518718be28/Snakefile#L18)

AndyMcAliley replied:
Done! This is nice.

jsadler2 commented on 1_fetch/src/sb_fetch.py:
```diff
@@ -0,0 +1,91 @@
+    response = sb_get(sb, item_id, filename, destination_dir=destination_dir)
+
+
+def parse_args():
```
If you use `script` instead of `shell`, Snakemake does all of the command line argument parsing for you:

* see simplified snakefile: https://github.com/jsadler2/lake-temperature-lstm-static/blob/62407e627e9feb079700d9c816c0cb13fc7d8b94/Snakefile#L32

* and simplified .py file: https://github.com/jsadler2/lake-temperature-lstm-static/blob/62407e627e9feb079700d9c816c0cb13fc7d8b94/1_fetch/src/sb_fetch.py#L69

AndyMcAliley replied:
Done! Using `script` does make things cleaner.

jsadler2 commented on Snakefile:
```diff
@@ -0,0 +1,37 @@
+configfile: "1_fetch/fetch_config.yaml"
```
if you do this, I would read the Snakemake docs about modularization (https://snakemake.readthedocs.io/en/stable/snakefiles/modularization.html). 

I think it would probably be simplest while also being sufficient to use the `include` funcationality where each phase's snakefile includes the relevant previous phases' snakefiles.

AndyMcAliley replied:
Done! I also added a new `Snakefile` which is the main file containing the rule `all`.

jsadler2 reviewed (commented):
Hey @AndyMcAliley, exciting to see this!

As we discussed on the call yesterday, I have a few suggestions and alternative ways you might think about using Snakemake for your pipeline.

Here are three different alternatives that you may find helpful.
* [restructuring config file and using expand to combine the suffixes and "driver types"](https://github.com/jsadler2/lake-temperature-lstm-static/tree/snakefile-shorter)
* [above + using input function to automatically populate the other input files](https://github.com/jsadler2/lake-temperature-lstm-static/blob/snakefile-even-shorter/Snakefile)
* [above + using `script` instead of `shell`](https://github.com/jsadler2/lake-temperature-lstm-static/tree/snakefile-script)

I would most suggest the four following adjustments:
1. use expand for the driver files with suffixes
2. add one more level to config so you can use just one rule
3. rename your snakemake file to 1_fetch.smk
4. use `script` instead of `shell` to save yourself some code/mental space

AndyMcAliley commented:
The workflow should be general enough to accommodate future data downloads from other Sciencebase items besides MNTOHA, or from sources other than Sciencebase. I renamed a number of things from "sb" to "mntoha" to try to make room for those future changes. In `1_fetch/out`, I also added `_mntoha` as a suffix to each folder in anticipation of pulling other data later. That suffix can help Snakemake figure out how to do the download.

hcorson-dosch commented on 1_fetch.smk:
```diff
@@ -0,0 +1,43 @@
+    input_files = []
+    # include files under files key in each category
+    for category in categories:
+        for filename in config["mntoha"][category]["files"]:
```
gosh I love dictionaries.

hcorson-dosch commented on 1_fetch.smk:
```diff
@@ -0,0 +1,43 @@
+    # include driver files: 3 types x 13 suffixes = 39 files
+    for driver_type in config["mntoha"]["drivers"]["driver_types"]:
+        for suffix in config["mntoha"]["drivers"]["suffixes"]:
+            input_files.append(os.path.join(out_dir, "drivers_mntoha", f"{driver_type}_{suffix}.zip"))
```
And f-strings! So much cleaner than `sprintf()` in R!

hcorson-dosch commented on 1_fetch.smk:
```diff
@@ -0,0 +1,43 @@
+    '''List all MNTOHA files to download
+    Includes metadata, temperature observations, and drivers
+
+    :param wildcards: Snakemake wildcards (unused)
```
Isn't `wildcards` used in line 40?

AndyMcAliley replied:
Line 40 is outside the scope of `get_mntoha_input_files`, so that `wildcards` is a different variable. We could rename one for clarity.

hcorson-dosch commented on 1_fetch.smk:
```diff
@@ -0,0 +1,43 @@
+    output:
+        "1_fetch/out/{file_category}_mntoha/{file}"
+    params:
+        sb_id = lambda wildcards: config["mntoha"][wildcards.file_category]["sb_id"]
```
Just trying to wrap my head around what `snakemake` is doing here. So this works b/c `snakemake` automatically uses the `mntoha` dictionary from `1_fetch/fetch_config.yml` as `wildcards` and then the keys from that dictionary as `file_category`?

AndyMcAliley replied:
Okay, here's my understanding. So, if the rule `fetch_mntoha_data_file` is being called, that means that `snakemake` has matched a requested file path to the pattern defined here in `output`. [`wildcards`](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#wildcards) contains both of the wildcards that were named in the pattern defined in `output`: `file_category` and `file`. If a function is called from within a `snakemake` rule, the first argument passed to that function is `wildcards`. That goes for both named functions (e.g. `get_mntoha_input_files`) and `lambda` functions like the one in line 40. That's why `wildcards` is named the same in lines 6 and 40 - it's that argument - even though in line 6, we know it's the `wildcards` from the `fetch_all` rule, while in line 40, it's the `wildcards` from the `fetch_mntoha_data_file` rule. From there, like you said, `snakemake` gets the `sb_id` by indexing the `config` that's loaded from `1_fetch/fetch_config.yaml`.

AndyMcAliley replied:
Alternatively, we could have defined a function, like:
```
def get_sb_id(wildcards):
    file_category = wildcards.file_category
    return config["mntoha"][file_category]["sb_id"]

rule fetch_mntoha_data_file:
    input:
        "1_fetch/in/pull_date.txt"
    output:
        "1_fetch/out/{file_category}_mntoha/{file}"
    params:
        sb_id = get_sb_id
    script:
        "1_fetch/src/sb_fetch.py"


```

hcorson-dosch replied:
Thanks, Andy - that makes complete sense now.

jsadler2 replied:
I agree with your explanation, @AndyMcAliley.

I'll add a concrete example, in case that is helpful. 

- Snakemake always works from what you are asking it to make. In `fetch_all`, one of the things we are asking it to make is `1_fetch/out/drivers/input_cells.png` (one file in the list of many files that the `get_mntoha_input_files` function returns).
- Snakemake finds the rule that makes `1_fetch/out/drivers/input_cells.png`. This is `fetch_mntoha_data_file` rule. 
- The `wildcards` object gets built by parsing the filename according to the wildcards defined in the `output` string of the rule (`"1_fetch/out/{file_category}_mntoha/{file}"`). 
- So in this case `wildcards.file_type = "drivers"` and `wildcards.file = "input_cells.png"`. 
- Then, the `params` are defined. In this case, the one-line `lambda` function will return `config["mntoha"]["drivers"]["sb_id"]` which is `"5e5d0b96e4b01d50924f2b34"` from the config file. 
- That `sb_id` is then accessed via the `params` object in the `1_fetch/src/sb_fetch.py` file since you are using the `script` functionality.

hcorson-dosch replied:
That is helpful -- thanks, Jeff!

hcorson-dosch commented on 1_fetch.smk:
```diff
@@ -0,0 +1,43 @@
+    input:
+        "1_fetch/in/pull_date.txt"
+    output:
+        "1_fetch/out/{file_category}_mntoha/{file}"
```
So `snakemake` knows this rule applies to all of the files returned by `get_mntoha_input_files` because they all match this  pattern, where {file_category} and {file} aren't actually known to `snakemake`, but it can infer them based on this pattern matching?

AndyMcAliley replied:
Right! Whenever a file path matches `1_fetch/out/*_mntoha/*`, `snakemake` calls this rule to make the file, where the wildcard `file_category` is the first `*` and `file` is the second `*`.

AndyMcAliley replied:
That's one thing about `snakemake` that I'm still wrapping my head around. It means that we should be careful about naming conventions, since file path matching is how `snakemake` determines how to use one rule to make multiple files.

hcorson-dosch commented on README.md:
````diff
@@ -1,2 +1,38 @@
+
+```
+conda update -n base conda
+conda env create -f environment.yaml
````
So nice to have the environment captured in a yaml file!

hcorson-dosch commented on README.md:
```diff
@@ -1,2 +1,38 @@
+
+`snakemake -c1 -p fetch_all`
+
+Replace the `1` in `-c1` with the number of cores to execute jobs in parallel.
```
So running jobs in parallel is that easy?! `snakemake` does it automatically if the job is allocated multiple cores?

AndyMcAliley replied:
I haven't tried this on HPC resources, but it worked on my laptop for the download!

hcorson-dosch commented on Snakefile:
```diff
@@ -0,0 +1,8 @@
+include: "1_fetch.smk"
```
So here's where rules associated with multiple phases of the code would be pulled together in to a master file, much like `_targets.R`?

AndyMcAliley replied:
Exactly. I believe if you call `snakemake` from the command line without specifying a file, it defaults to `Snakefile`.

hcorson-dosch commented on 1_fetch/src/sb_fetch.py:
```diff
@@ -0,0 +1,70 @@
+
+
+if __name__ == '__main__':
+    main(snakemake.params['sb_id'], snakemake.output[0])
```
Hmm I'm a bit confused -- if no files were listed under a given category (in this example, `metadata`) in the `mntoha` dictionary defined in `1_fetch/fetch_config.yaml`, like so:
```
mntoha:

  metadata:
      sb_id: 5e5c1c1ce4b01d50924f27e7
      files:

  obs:
      sb_id: 5e5d0b68e4b01d50924f2b32
      files:
          - temperature_observations.zip
```
Then wouldn't no files (nor even a directory) be added for that category by the `get_mntoha_input_files()` function, which would then mean that the `fetch_mntoha_data_file` rule wouldn't identify `metadata` as a `file_category`? And therefore the `1_fetch/src/sb_fetch.py` script wouldn't be called for that category?

Just trying to think through when the `filepath = None` default argument would be used, and I'm not sure how it would work given this `snakemake` setup. But I may be missing something.

AndyMcAliley replied:
You're right - the way it's set up now, `filepath=None` wouldn't be used at any point, and that code in `sb_fetch` doesn't make sense. I wrote that code when I was using `snakemake` in a different way, using `shell` instead of `script`. I can remove the logic for `filepath=None` from `main`.

hcorson-dosch commented on 1_fetch/src/sb_fetch.py:
```diff
@@ -0,0 +1,70 @@
+        destination_dir = os.path.dirname(filepath)
+        if not os.path.exists(destination_dir):
+            os.makedirs(destination_dir)
+        if filename == '':
```
Similarly to my comment on line 70 - when would this be the case, given the current setup? if you provided this, with just a space or nothing after the hash?
```
mntoha:

  metadata:
      sb_id: 5e5c1c1ce4b01d50924f27e7
      files:
         -  
  obs:
      sb_id: 5e5d0b68e4b01d50924f2b32
      files:
          - temperature_observations.zip
```
And if that is simply a space (or an empty string) for the `['metadata']['files']` value in the `mntoha` dictionary, then does `snakemake` match `''`  for `{file}` in the wildcard using the `"1_fetch/out/{file_category}_mntoha/{file}"` pattern?

AndyMcAliley replied:
Right, as above, that logic won't get triggered by `snakemake` either.

AndyMcAliley replied:
If we wanted to download all the files in one SB item... we could rewrite `main` in `sb_fetch.py`, and probably include a boolean in Snakemake.params that signifies whether we want the single files or all files in the SB item. Do you think we'll want that functionality at some point?

AndyMcAliley replied:
(I tried a few iterations of None/empty string for `['metadata']['files']` and got errors)

hcorson-dosch replied:
I think that leaving as is (having to specify the files) is fine for now, given the current use case, which is the mntoha data. If we switch to a new data sources, that might either 1) motivate adding the option to download all files from a given SB id or 2) require changes to this framework anyhow (e.g., a total hypothetical, but maybe we decide to store data on caldera instead and pull with globus sdk)

hcorson-dosch reviewed (approved):
Thanks for answering my many questions, Andy. This all looks great to  me.
